lib/pbench/agent/redis.py

```diff
@@ -222,39 +222,83 @@ def emit(self, record):
+
+    Arguments:
+
+        redis_server: a Redis client object
+        key: the key name as a string
+        prog: the program name of our caller
+
+    Returns the payload value of the key as a string.
+
+    If successful on the first connection attempt, and the value of the key is
+    retrieved, no messages are issued on stdout.  If the key does not exist yet,
+    we'll periodically report the missing key from the last successful
+    connection.
+
+    If on the first connection attempt the connection to the Redis server fails,
+    we emit a message to that effect, once.  After that, we'll report each time
+    we are disconnected or re-connected to the Redis server.
     """
     # Loop waiting for the key to show up.
     connected = None
     payload = None
-    while payload is None:
+    attempts = 0
+    errors = 0
+    missing = 0
+    while True:
+        attempts += 1
         try:
             payload = redis_server.get(key)
         except redis.ConnectionError:
+            errors += 1
+            # Reset the missing count since we got disconnected
+            missing = 0
             if connected is None:
+                # Only emit this message once for the initial attempt.
                 print(
-                    f"{prog}: waiting to connect to redis server {redis_host}:{redis_port}",
+                    f"{prog}: waiting to connect to Redis server {redis_server}",
                     flush=True,
                 )
-                connected = False
             elif connected:
+                # We always report disconnections.
                 print(
-                    f"{prog}: disconnected from redis server {redis_host}:{redis_port}",
+                    f"{prog}: disconnected from Redis server {redis_server}",
                     flush=True,
                 )
-                connected = False
-            time.sleep(1)
+            connected = False
         else:
-            if not connected:
+            if connected is False:
+                # We always report re-connections (connections was not None).
                 print(
-                    f"{prog}: connected to redis server {redis_host}:{redis_port}",
-                    flush=True,
+                    f"{prog}: connected to Redis server {redis_server}", flush=True,
                 )
-                connected = True
-            if payload is None:
-                print(f'{prog}: key, "{key}" does not exist yet', flush=True)
-                time.sleep(1)
+            if payload is not None:
+                break
+            connected = True
+            if (missing % _MISSING_KEY_REPORT_INTERVAL) == 0:
+                # Only emit missing key notice periodically
+                print(f"{prog}: no payload for key '{key}', yet", flush=True)
+            missing += 1
+        time.sleep(_CONNECTION_RETRY_INTERVAL)
+
+    if attempts > 1:
+        # Always report stats if it took multiple attempts.
+        print(
```

Comment:
So, should these messages be _printed_?...or should they be sent to some sort of logger which might direct them to the terminal?

That is, code might be called via an API from a second-party application, in which case they might not appreciate random messages being spit out....

Reply:
Really good question.

The two entities using this are the Tool Meister and the Tool Data Sink.  There are two modes of operation of those services today: containerized or non-containerized.

In the non-containerized case we "daemonize" to run in the background, but we only setup the logging after we daemonize to avoid inadvertently closing the logging `fd`.

So we use `print()`.  We could just hand a function here instead so that they could use a logger.